### PR TITLE
fix mypy typing errors in pytorch_lightning/strategies/dp.py

### PR DESCRIPTION
**Pull request by Cyprien-Ricque**
## What does this PR do?

Fixes mypy typing errors in pytorch_lightning/strategies/dp.py in https://github.com/Lightning-AI/lightning/issues/13445.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [X] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by otaj — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
lgtm after the comment from @otaj

File: src/pytorch_lightning/strategies/dp.py [hunk 1]
```diff
@@ -67,6 +72,9 @@ def world_size(self) -> int:
     def setup(self, trainer: "pl.Trainer") -> None:
         # model needs to be moved to the device before it is wrapped
         self.model_to_device()
+        assert isinstance(self.model, pl.LightningModule) or isinstance(
+            self.model, _LightningPrecisionModuleWrapperBase
+        )
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/dp.py:77]:
```suggestion
        assert isinstance(self.model, (pl.LightningModule, _LightningPrecisionModuleWrapperBase))
```

**Review by rohitgr7 — APPROVED** [review]

File: src/pytorch_lightning/strategies/dp.py [hunk 2]
```diff
@@ -11,21 +11,26 @@
...
 from pytorch_lightning.plugins.io.checkpoint_plugin import CheckpointIO
 from pytorch_lightning.plugins.precision import PrecisionPlugin
 from pytorch_lightning.strategies.parallel import ParallelStrategy
+from pytorch_lightning.strategies.strategy import TBroadcast
 from pytorch_lightning.utilities.apply_func import apply_to_collection
+from pytorch_lightning.utilities.distributed import ReduceOp
 from pytorch_lightning.utilities.model_helpers import is_overridden
 from pytorch_lightning.utilities.types import _METRIC_COLLECTION, STEP_OUTPUT
 
+T = TypeVar("T")  # Both input and return type of DataParallelStrategy.reduce()
```
[the remark this answers is not in the document]
Comment (carmocca) [reply src/pytorch_lightning/strategies/dp.py:32]: `TReduce` would be consistent with `TBroadcast`. Perhaps they should be defined together

[the remark this answers is not in the document]
Comment (otaj) [reply src/pytorch_lightning/strategies/dp.py]: Resolved

**Pull request merged**